### ISSUE TITLE
Unable to produce to AWS MSK from AWS Lambda

### ISSUE THREAD

**Issue by scimas** [fede1024/rust-rdkafka#513]
I'm trying to produce some messages to a topic in MSK and having issues. I have verified that this is a not a networking issue by creating a python replica of the lambda - with the same VPC and security group configuration; and the python code is able to produce to the topic without any issues.

All I've received from the rdkafka logs is as follows. There are no errors, the lambda just times out after 1 minute. Increasing the timeout does not help. Any suggestion on how to proceed from here would be appreciated.
```
2022-11-07T20:37:37.920Z TRACE [rdkafka::client] Create new librdkafka client 0x55a7306ea0
2022-11-07T20:37:37.920Z INFO  [rust_testing] sending message
2022-11-07T20:37:37.920Z TRACE [rdkafka::producer::base_producer] Polling thread loop started
```

The rust code:
```rs
use std::time::Duration;
use rdkafka::{producer::{FutureProducer, FutureRecord, Producer}, util::Timeout};
use lambda_runtime::LambdaEvent;

#[tokio::main]
async fn main() -> Result<(), lambda_runtime::Error> {
    simple_logger::init_with_level(log::Level::Trace)?;
    let lambda_fn = lambda_runtime::service_fn(lambda_handler);
    lambda_runtime::run(lambda_fn).await?;
    Ok(())
}

async fn lambda_handler(_payload: LambdaEvent<serde_json::Value>) -> Result<(), lambda_runtime::Error> {
    let topic = "my-topic";
    let bootstrap_servers = "server1:9092,server2:9092".to_string();
    let producer: FutureProducer = rdkafka::config::ClientConfig::new()
        .set("bootstrap.servers", bootstrap_servers)
        .set("security.protocol", "plaintext")
        .create()?;
    let payload = r#"{"key": "myval"}"#.to_string();
    log::info!("sending message");
    match producer.send(FutureRecord::<String, _>::to(&topic).payload(&payload), Timeout::After(Duration::from_secs(15))).await {
        Ok(_) => (),
        Err(e) => println!("error: {e:?}"),
    }
    Ok(())
}
```

And the python code that works:
```py
import json
from kafka import KafkaProducer

def lambda_handler(event, context):
    topic = "my-topic"
    bootstrap_servers = "server1:9092,server2:9092"
    producer = KafkaProducer(bootstrap_servers=bootstrap_servers, value_serializer=lambda v: json.dumps(v).encode("utf-8"))
    data = {"key": "myval"}
    producer.send(topic, data)
    producer.flush()
    producer.close()
```

**Comment by scimas** [comment]
I ran the same test case with the [kafka crate](https://crates.io/crates/kafka) and it works. So I'm pretty sure this issue is with rdkafka and/or the configuration of the client. Again, any hint of how to proceed would be nice, since the kafka crate does not seem to be maintained very actively.

**Comment by gklijs** [comment]
The result is a future, so you will also need to await that. Like [here](https://github.com/fede1024/rust-rdkafka/blob/35ad5ba04ca5ecad73a3e7569b5c66d1126bcdf4/examples/simple_producer.rs#L48). Currently, it probably quits before really sending the record.

**Comment by scimas** [comment]
The result is not a future, as far as I can tell. `FutureProducer.send` is a future, which I'm awaiting in the `match`. `FutureProducer.send(...).await` returns an [`OwnedDeliveryResult`](https://docs.rs/rdkafka/0.29.0/rdkafka/producer/future_producer/struct.FutureProducer.html#method.send), which is just a [`Result`](https://docs.rs/rdkafka/0.29.0/rdkafka/producer/future_producer/type.OwnedDeliveryResult.html), I don't see how I can `.await` it. And that's why I have the
```rust
match producer.send(...).await {
    Ok(_) => (),
    Err(e) => println,
}
```
block in the first place.

**Comment by gklijs** [comment]
Fair enough, you could log the value of the [result](https://docs.rs/rdkafka/0.29.0/rdkafka/producer/future_producer/type.OwnedDeliveryResult.html) though, that hopefully points to the right direction.

**Comment by scimas** [comment]
There is _no result_ to log, that's the issue. The `.await` never resolves. Otherwise the function would end either in an error through the `Err(e) => println` path or the "do nothing" `Ok(_) => ()` path.

**Comment by scimas** [comment]
So it is looking likely that this is an issue with `librdkafka` - AWS MSK / AWS Lambda.

I tried the same test as my original post with python again. The first test I had stated was done with the kafka-python package, which is a python-native implementation of a kafka client. Today I tried a test with the confluent-kafka package which, like rust-rdkakfa, uses librdkafka. This test failed. Same behavior where the call to `Producer.flush()` never returns.

I will be reporting this at the librdkafka repo, referencing this issue. I'll leave it up to you whether to keep this issue open or not.

I am still at a loss as to what to report other than "it works without librdkafka, but doesn't with it."

**Comment by benesch** [comment]
Recommend that you enable debug logs, via `config.set("debug", "all")`. You should also make sure you have `env_logger` or such installed, so that the log messages actually get printed to stderr.

**Comment by scimas** [comment]
Solved.

Issue was that the broker had `min.insync.replicas=2`, the replication factor of the topic I was testing with was 1 and librdkafka requires `acks=all` by default. So it kept failing and silently retrying.

Two possible solutions
1. Set `acks` explicitly to be equal to `min(replication factor, min.insync.replicas)`.
2. Set `message.timeout.ms` to a value acceptable for you. By default it is 5 min, which is way longer than I was waiting (1 min). If this was smaller than 1 min, it might have actually returned an error.

I went with 1 for testing purposes, probably will go for 2 in actual use.
